### PR TITLE
Ade 582 create rest endpoints and admin page

### PR DESCRIPTION
**Pull request by shouvSarker**
PRing on Robin's branch. Have not squashed commits for ease of reviewing, commits will be squashed to present organised commit message when PRing to develop.

### CODE REVIEW

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java [hunk 1]
```diff
@@ -107,4 +149,131 @@ public static MacroRepresentation getMacroRepresentation(NodeList macros, Struct
...
+   * @param serializer XML serialiser
+   * @return Representation of a macro from Confluence Storage format
+   */
+  private static MacroRepresentation getMacroFromList(
+    NodeList macros, StructuredCategory category, LSSerializer serializer) {
+    return getMacroRepresentation(macros, category, serializer, xhtmlContent);
+  }
+
+  /**
+   *
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: What is the purpose of this method? Is "Utilities" the best place for it to live?

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java [hunk 2]
```diff
@@ -107,4 +149,131 @@ public static MacroRepresentation getMacroRepresentation(NodeList macros, Struct
...
+    SearchFilter searchFilter =
+      ContentPermissionsSearchFilter.getInstance()
+        .and(SpacePermissionsSearchFilter.getInstance());
+
+    ContentSearch search =
+      new ContentSearch(
+        createSearchableQuery(labels, spaceKey),
+        ModifiedSort.DESCENDING,
+        searchFilter,
+        new SubsetResultFilter(100));
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: What impact does this magic number have?

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java [hunk 3]
```diff
@@ -107,4 +149,131 @@ public static MacroRepresentation getMacroRepresentation(NodeList macros, Struct
...
+    List<IdeaContainer> rows = Collections.emptyList();
+
+    try {
+      List<Searchable> maybePages =
+        searchManager.searchEntities(search, EntityVersionPolicy.LATEST_VERSION);
+
+      rows =
+        maybePages.stream()
+          .filter(entity -> entity instanceof AbstractPage)
+          .map(
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: I would be inclined to combine the `filter` and `map` into a single `flatMap` that returns nothing if the `instanceof` check fails.

Reason being that these two steps currently rely on each other, better to have that knowledge in one place.

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java [hunk 4]
```diff
@@ -107,4 +149,131 @@ public static MacroRepresentation getMacroRepresentation(NodeList macros, Struct
...
+            entity -> {
+              AbstractPage page = (AbstractPage) entity;
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: Chop this transformation out into a separate method.

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java [hunk 5]
```diff
@@ -107,4 +149,131 @@ public static MacroRepresentation getMacroRepresentation(NodeList macros, Struct
...
+
+                Arrays.asList(StructuredCategory.values())
+                  .forEach(
+                    category ->
+                      row.setMacroRepresentations(
+                        category, getMacroFromList(macros, category, serializer)));
+
+                return row;
+              } catch (ParserConfigurationException | IOException | SAXException e) {
+                log.warn(e.toString());
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: What is the difference between logging the string representation of an exception and the exception itself? 

Generally, it is best to add context and action taken when logging exceptions. e.g "Could not X the Y, ignoring".

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java [hunk 6]
```diff
@@ -107,4 +149,131 @@ public static MacroRepresentation getMacroRepresentation(NodeList macros, Struct
...
+              }
+
+              return null;
+            })
+          .filter(Objects::nonNull)
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: See previous comment, `flatMap` is likely better.

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java [hunk 7]
```diff
@@ -107,4 +149,131 @@ public static MacroRepresentation getMacroRepresentation(NodeList macros, Struct
...
+          .collect(Collectors.toList());
+    } catch (InvalidSearchException e) {
+      log.warn(e.toString());
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: See previous comment. When would this happen? Only due to programmer error?

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java [hunk 8]
```diff
@@ -1,60 +1,54 @@
...
+import org.springframework.stereotype.Component;
 
 /**
  * External rest API servlet
  */
 @Path("/")
 @Component
-public class TechnologyList {
-  private static final Logger log = LoggerFactory.getLogger(TechnologyList.class);
+public class index {
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Lowercase `i` here is unusual. What drove that decision?

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java [hunk 9]
```diff
@@ -197,7 +187,101 @@ public String getSchemaIds( @Context HttpServletResponse response){
     return this.gson.toJson( schemaIds);
   }
 
+  /**
+   * @param searchString to find technologies that begin with this string
+   * @param response     Servlet contest
+   * @return String in the form of a json list of TechnologyAPI objects
+   */
+  @Path("/ideaPages")
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Mixed case is unusual.

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java [hunk 10]
```diff
@@ -197,7 +187,101 @@ public String getSchemaIds( @Context HttpServletResponse response){
     return this.gson.toJson( schemaIds);
   }
 
+  /**
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: What does this do?

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java [hunk 11]
```diff
@@ -197,7 +187,101 @@ public String getSchemaIds( @Context HttpServletResponse response){
...
+
+      preJsonIdea.put("owner", idea.getOwner());
+
+      preJsonIdea.put("status", idea.getStatus());
+
+      return preJsonIdea;
+    }).collect(Collectors.toList());
+
+    return allIdeas.isEmpty()
+      ? "[{}]"
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Oof. Why is this necessary?

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java [hunk 12]
```diff
@@ -197,7 +187,101 @@ public String getSchemaIds( @Context HttpServletResponse response){
...
+
+    title = title != null ? title : "";
+    description = description != null ? description : "";
+    status = status != null ? status : "";
+    owner = owner != null ? owner : "";
+
+    List<FedexIdea> allIdeas = this.fedexIdeaService.queryAllFedexIdea(title, description, status, owner);
+
+    List<Map> preConvert = allIdeas.stream().map( idea -> {
+      Map preJsonIdea = new HashMap<String, String>();
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: This looks like a mapping to a DTO. This should be automated using the index schema?

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java [hunk 13]
```diff
@@ -197,7 +187,101 @@ public String getSchemaIds( @Context HttpServletResponse response){
...
+    @Context HttpServletRequest request,
+    @Context HttpServletResponse response
+  ) {
+    this.applyNoCacheHeaders(response);
+    String schemaBody = "";
+
+    try {
+      schemaBody = extractPostRequestBody(request);
+    } catch (Exception e){
+      throw new Error("Error parsing request body");
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: `extractPostRequestBody` already swallows all exceptions and returns the string representation of the exception (for some reason) - what is this trying to do?

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java [hunk 14]
```diff
@@ -197,7 +187,101 @@ public String getSchemaIds( @Context HttpServletResponse response){
...
+    }
+
+    Map mappedSchemaBody = this.gson.fromJson(schemaBody, Map.class);
+
+    List<FedexSchema> allSchema = this.fedexIdeaService.listSchemas();
+    FedexSchema latestSchema = allSchema.get(allSchema.size() - 1);
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: What if there are no schemas?

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java [hunk 15]
```diff
@@ -197,7 +187,101 @@ public String getSchemaIds( @Context HttpServletResponse response){
...
+
+    latestSchema.setIndexSchema(mappedSchemaBody.get("indexSchema").toString());
+
+    latestSchema.setUiSchema(mappedSchemaBody.get("uiSchema").toString());
+
+    latestSchema.setSchema(mappedSchemaBody.get("schema").toString());
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: This sort of mapping should be done with something like https://mapstruct.org/

File: frontend-react/src/components/AdminPage.tsx [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+  return errors;
+};
+
+const OuterAdminForm = () => {
+  const contextPath = window.AJS ? window.AJS.contextPath() : "/confluence";
+
+  const [formData, setFormData] = useState<formDataType>();
+
+  useEffect(() => {
+    axios.get(`${contextPath}/rest/idea/1/schema`).then((response) =>
```
Comment (dspasojevic) [inline frontend-react/src/components/AdminPage.tsx]: What does `1` mean here?

File: frontend-react/src/components/AdminPage.tsx [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+        onSubmit={() => {
+          updateSchema(formData);
+        }}
+      >
+        <Button type="submit" appearance="primary">
+          Save
+        </Button>
+        <div hidden={submissionFeedback.hidden}>
+          <SectionMessage
+            title={submissionFeedback.title} //"This account will be permanently deleted"
```
Comment (dspasojevic) [inline frontend-react/src/components/AdminPage.tsx]: What do these comments mean?

File: frontend-react/src/components/Table.tsx [hunk 18]
```diff
@@ -1,190 +1,155 @@
...
   };
 
-
-
+  const [justPages, setJustPages] = useState<Array<IdeaPage>>();
+
+  useEffect(() => {
+    axios
+      .get(
+        `${contextPath}/rest/idea/1/ideaPages?description=` +
```
Comment (dspasojevic) [inline frontend-react/src/components/Table.tsx]: `1`?

Is there a way to construct this without string concatenation?

**Comment by dspasojevic** [comment]
I'm surprised by the number of times idea specific things are happening. When do they start to go away.

In general, code should be commented and thoughtfully written. When naming methods, prefer to reserve 'verbs' unless the language / framework dictates otherwise (e.g. getters and setters) or if they are side-effecting / mutating.

File: frontend-react/src/components/AdminPage.tsx (hunk 16 above)
Comment (robinMcA) [reply frontend-react/src/components/AdminPage.tsx]: From memory it is the version of the api.
`<rest key="rest" name="Idea Rest API" path="/idea" version="1">`
would it be better as v1?

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java (hunk 1 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: The method was curved out of IndexTable.java to provide a common function to get rows. We do not need it anymore as we get the index pages from our rest endpoint. Removing it.

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java (hunk 2 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: It was searching for up to 100 pages. Function is removed now.

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 8 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: It was done by mistake while renaming the file. Rectifying it.

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 9 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Updated it

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java (hunk 7 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: Function removed. InvalidSearchException could occur from searchManager, not handling it gives unhandled exception error.

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java (hunk 6 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: Function removed

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java (hunk 5 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: Function removed. Will be mindful about logging exceptions with more meaningful messages from next time.

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java (hunk 4 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: Function removed

File: backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java (hunk 3 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/helpers/Utilities.java]: Function removed.

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 10 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Updated documentation to reflect this function's purpose

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 13 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Updated to throw error instead of returning a string representation of it. That was the original intention.

File: frontend-react/src/components/AdminPage.tsx (hunk 17 above)
Comment (shouvSarker) [reply frontend-react/src/components/AdminPage.tsx]: Removed the comments

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 13 above)
Comment (dspasojevic) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: By doing this, you've hidden information about the original exception. Consider something like:
```
throw new Exception("Failed to parse request body.", e)
```

Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Updated!

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 12 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: We are not using schemas yet, but it should be automated when index schema is used to decide which fields are shown on index page

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 11 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: It's not necessary anymore. Removed it.

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 14 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: It will give out a 500 error as it will crash on this line. Have put in a check and initialising a schema if there is none.

File: frontend-react/src/components/Table.tsx (hunk 18 above)
Comment (shouvSarker) [reply frontend-react/src/components/Table.tsx]: Updated to use string interpolation and query-search instead of manually constructing search string.

File: frontend-react/src/components/AdminPage.tsx (hunk 16 above)
Comment (shouvSarker) [reply frontend-react/src/components/AdminPage.tsx]: It seems like atlassian doesn't recognise `v1` as a valid version

`The version (v1)set on the REST module descriptor 'au.com.agiledigital.idea_search:rest (Provides a REST interface to expose and modify data.)//rest' of plugin 'au.com.agiledigital.idea_search' is not valid. It must follow the following pattern 'major[.minor][.micro][.classifier]'`

Comment (shouvSarker) [reply frontend-react/src/components/AdminPage.tsx]: Declared version as a const and then using it instead of using `1` here.

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 13 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Updated again to avoid rethrowing.

File: backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java (hunk 15 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/rest/index.java]: Updated to do mapping with mapstruct!

**Comment by shouvSarker** [comment]
> I'm surprised by the number of times idea specific things are happening. When do they start to go away.
> 
> In general, code should be commented and thoughtfully written. When naming methods, prefer to reserve 'verbs' unless the language / framework dictates otherwise (e.g. getters and setters) or if they are side-effecting / mutating.

Hopefully, we will see less of idea and more generics happening after ADE-613, 604 and 632 goes through. 

I will be more mindful about commenting, conventions and tooling (for example, using mapstruct instead of manually doing dto) in future.
